File: 113.路径总和-ii.js
/*
 * @lc app=leetcode.cn id=113 lang=javascript
 *
 * [113] 路径总和 II
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number[][]}
 */
var pathSum = function(root, targetSum) {

    let res = [];

    let backtrack = function(node, path, sum, target) {
        if (!node) {
            return;
        }

        path.push(node.val);
        sum += node.val;

        if (sum === target && !node.left && !node.right) {
            res.push([...path])
        }

        backtrack(node.left, path, sum, target);
        backtrack(node.right, path, sum, target);

        sum -= node.val;
        path.pop();
    }

    backtrack(root, [], 0, targetSum);

    return res;

};
// @lc code=end

File: 113.路径总和-ii.java
/*
 * @lc app=leetcode.cn id=113 lang=java
 *
 * [113] 路径总和 II
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    ArrayList<List<Integer>> res = new ArrayList<List<Integer>>();

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {

        backtrack(root, new ArrayList<Integer>(), 0, targetSum);

        return res;
    }

    private void backtrack(TreeNode node, ArrayList<Integer> path, int sum, int targetSum) {

        if (node == null) {
            return;
        }

        path.add(node.val);
        sum += node.val;

        if (sum == targetSum && node.left == null && node.right == null) {
            res.add(new ArrayList<Integer>(path));
        }

        backtrack(node.left, path, sum, targetSum);
        backtrack(node.right, path, sum, targetSum);

        sum -= node.val;
        path.remove(path.size() - 1);
    }
}
// @lc code=end

